File: 20-BinarySearchProblems/78-subsets/subsets.cpp
class Solution {
public:
    void helper( int i, vector<int>& nums, vector<int>& subset, vector<vector<int>>& ans){

        int n = nums.size();
        if( i == n ){
            ans.push_back(subset);
            return;
        }
        subset.push_back(nums[i]);
        helper( i+1, nums, subset, ans);
        
        subset.pop_back();
        helper( i+1, nums, subset, ans);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans; 

        vector<int> subset ; 
        helper(0, nums, subset, ans);
        return ans;
    }
};

File: 20-BinarySearchProblems/78-subsets/subsets.java
class Solution {
    public static void solve(List<List<Integer>>res, int[] nums,List<Integer>curr,int i) {
        if(i == nums.length) {
            System.out.println("base case: "+curr);
            res.add(new ArrayList<>(curr));
            return;
        }

        curr.add(nums[i]);
        System.out.println("Before: " +curr);
        solve(res,nums,curr,i+1);
        System.out.println("After 1: " +curr);
        curr.remove(curr.size()-1);
        solve(res,nums,curr,i+1);
        System.out.println("After 2: " +curr);
    }
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer>curr = new ArrayList<>();

        solve(res,nums,curr,0);
        return res;
    }
}